parquet/src/main/java/org/apache/iceberg/parquet/ParquetBloomRowGroupFilter.java

```diff
@@ -0,0 +1,330 @@
+      // bloom filter only contain non-nulls and cannot eliminate based on isNull or NotNull
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean isNaN(BoundReference<T> ref) {
+      // bloom filter is based on hash and cannot eliminate based on isNaN or notNaN
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean notNaN(BoundReference<T> ref) {
+      // bloom filter is based on hash and cannot eliminate based on isNaN or notNaN
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean lt(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean ltEq(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gt(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean gtEq(BoundReference<T> ref, Literal<T> lit) {
+      // bloom filter is based on hash and cannot eliminate based on lt or ltEq or gt or gtEq
+      return ROWS_MIGHT_MATCH;
+    }
+
+    @Override
+    public <T> Boolean eq(BoundReference<T> ref, Literal<T> lit) {
+      int id = ref.fieldId();
+      ColumnDescriptor col = cols.get(id);
+      if (col == null) {
+        return ROWS_MIGHT_MATCH;
+      } else {
+        Boolean hasNoBloomFilter = fieldsWithNoBloom.contains(id);
+        if (hasNoBloomFilter) {
+          return ROWS_MIGHT_MATCH;
+        }
+        BloomFilter bloom = getBloomById(id);
+        Type type = types.get(id);
+        T value = lit.value();
+        return shouldRead(col, value, bloom, type);
+      }
+    }
+
+    private BloomFilter getBloomById(int id) {
+      if (bloomCache == null) {
+        bloomCache = Maps.newHashMap();
+      }
+      if (bloomCache.containsKey(id)) {
+        return bloomCache.get(id);
+      } else {
+        ColumnChunkMetaData columnChunkMetaData = columnMetaMap.get(id);
+        BloomFilter bloomFilter = bloomReader.readBloomFilter(columnChunkMetaData);
+        if (bloomFilter == null) {
+          throw new IllegalStateException("Failed to read required bloom filter for id: " + id);
+        } else {
+          bloomCache.put(id, bloomFilter);
+        }
+
+        return bloomFilter;
+      }
+    }
+
+    private <T> boolean shouldRead(ColumnDescriptor col, T value, BloomFilter bloom, Type type) {
+      long hashValue = 0;
+      switch (type.typeId()) {
+        case INTEGER:
+        case LONG:
+        case FLOAT:
+        case DOUBLE:
+        case DATE:
+        case TIME:
+        case TIMESTAMP:
+          hashValue = bloom.hash(value);
```

Comment:
Why does this not cast `value`? As long as the type is checked in this method, I think it makes sense to call the right `hash` method, instead of calling a generic one for `Object`.

Reply:
Casted. Thanks!